api/src/middleware/error-handler.ts

```diff
@@ -1,105 +1,118 @@
 			logger.debug(error);
 
-			if (!status) {
+			if (status === null) {
+				// Use current error status as response status
 				status = error.status;
 			} else if (status !== error.status) {
-				status = 500;
+				// Fallback if status has already been set by a preceding error
+				// and doesn't match the current one
+				status = FALLBACK_ERROR.status;
 			}
 
-			payload.errors.push({
+			errors.push({
 				message: error.message,
 				extensions: {
+					// Expose error code under error's extensions data
 					code: error.code,
 					...(error.extensions ?? {}),
 				},
 			});
 
 			if (isDirectusError(error, ErrorCode.MethodNotAllowed)) {
-				res.header('Allow', (error as InstanceType<typeof MethodNotAllowedError>).extensions.allowed.join(', '));
+				res.header('Allow', error.extensions.allowed.join(', '));
 			}
 		} else {
 			logger.error(error);
 
-			status = 500;
+			status = FALLBACK_ERROR.status;
 
 			if (req.accountability?.admin === true) {
 				const localError = isObject(error) ? error : {};
-				const message = localError['message'] ?? typeof error === 'string' ? error : null;
-
-				payload = {
-					errors: [
-						{
-							message: message || 'An unexpected error occurred.',
-							extensions: {
-								code: 'INTERNAL_SERVER_ERROR',
-								...(localError['extensions'] ?? {}),
-							},
+
+				// Use 'message' prop if available, otherwise if 'error' is a string use that
+				const message =
+					(typeof localError['message'] === 'string' ? localError['message'] : null) ??
+					(typeof error === 'string' ? error : null);
+
+				errors = [
+					{
+						message: message || FALLBACK_ERROR.message,
+						extensions: {
+							code: FALLBACK_ERROR.code,
+							...(localError['extensions'] ?? {}),
 						},
-					],
-				};
+					},
+				];
 			} else {
-				payload = {
-					errors: [
-						{
-							message: 'An unexpected error occurred.',
-							extensions: {
-								code: 'INTERNAL_SERVER_ERROR',
-							},
-						},
-					],
-				};
+				// Don't expose unknown errors to non-admin users
+				errors = [{ message: FALLBACK_ERROR.message, extensions: { code: FALLBACK_ERROR.code } }];
 			}
 		}
 	}
 
-	res.status(status ?? 500);
-
-	emitter
-		.emitFilter(
-			'request.error',
-			payload.errors,
-			{},
-			{
-				database: getDatabase(),
-				schema: req.schema,
-				accountability: req.accountability ?? null,
-			},
-		)
-		.then((updatedErrors) => {
-			return res.json({ ...payload, errors: updatedErrors });
-		});
-};
+	res.status(status ?? FALLBACK_ERROR.status);
+
+	const updatedErrors = await emitter.emitFilter(
+		'request.error',
+		errors,
+		{},
+		{
+			database: getDatabase(),
+			schema: req.schema,
+			accountability: req.accountability ?? null,
+		},
+	);
 
-export default errorHandler;
+	return res.json({ errors: updatedErrors });
+});
+
+function asyncErrorHandler(fn: ErrorRequestHandler) {
+	return (err: any, req: Request, res: Response, next: NextFunction) =>
```

Comment:
```suggestion
    // Note: keep all 4 parameters here. That's how Express recognizes it's the error handler, even if
    // we don't use next
	return (err: any, req: Request, res: Response, next: NextFunction) =>
```

Lets not forget adding back the original comment here

Reply:
Covered via tests now 👍